File: java_api/src/main/java/globalincidents/controller/GetGroupedCountries.java
package globalincidents.controller;

import org.json.JSONArray;
import org.json.JSONObject;
import org.springframework.util.MultiValueMap;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;
import utils.DBConnection;

import java.util.Collections;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

@RequestMapping(value = "/grouped-countries")
@RestController
public class GetGroupedCountries extends BaseController {
    static final int MAX_THREADS_NO = 5;

    /**
     * Challenge: Why is this call returning different values everytime time?
     */
    @Override
    public String run(@RequestParam MultiValueMap<String,String> params) {
        Map<String, Integer> toReturn = Collections.synchronizedMap(new HashMap<String, Integer>());

        JSONArray results = DBConnection.ExecuteQuery("select id, title from incidents");
        int limit = results.length() / MAX_THREADS_NO;

        ExecutorService executor = Executors.newFixedThreadPool(MAX_THREADS_NO);
        for (int i = 0; i < MAX_THREADS_NO; i++) {
            int index = i;
            Thread worker = new Thread(() -> {
                for( int jsonObjIndex = index * limit; jsonObjIndex < limit * (index + 1); jsonObjIndex++ ) {
                    String title = results.getJSONObject(jsonObjIndex).getString("title");
                    String[] splittedTitle = title.split(",");
                    String country = splittedTitle[splittedTitle.length - 1];


                    int currentValue = toReturn.getOrDefault(country, 0);
                    toReturn.put(country, currentValue + 1);
                }
            });
            executor.execute(worker);
        }
        executor.shutdown();
        while (!executor.isTerminated()) {
            try {
                Thread.sleep(100);
            } catch(Exception e){
            }
        }

        JSONObject json = new JSONObject(toReturn);
        return json.toString();
    }

    @Override
    public void getParams(MultiValueMap<String, String> params) {
    }

}

File: kotlin_api/src/main/kotlin/globalincidents/controller/GetGroupedCountries.kt
package globalincidents.controller

import org.json.JSONObject
import org.springframework.util.MultiValueMap
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController
import utils.DBConnection
import java.util.*
import java.util.concurrent.Executors


@RestController
@RequestMapping("/grouped-countries")
class GetGroupedCountries: BaseController(){
    companion object {
        const val MAX_THREADS_NO = 5
    }

    /**
     * Challenge: Why is this call returning different values everytime time?
     */
    override fun run(@RequestParam params: MultiValueMap<String, String>): String {
        val toReturn = Collections.synchronizedMap(HashMap<String, Int>())
        val results = DBConnection.executeQuery("select id, title from incidents")
        val limit = results.length() / MAX_THREADS_NO

        val executor = Executors.newFixedThreadPool(MAX_THREADS_NO)
        for (i in 0..(MAX_THREADS_NO - 1)) {
            val worker = Runnable {
                for( jsonObjIndex in (i * limit) until (limit * (i + 1))) {

                    val description = results.getJSONObject(jsonObjIndex).getString("title")
                    val country = description.split(",").last()
                    val currentValue = toReturn.getOrDefault(country, 0)
                    toReturn[country] = currentValue + 1
                }
            }
            executor.execute(worker)
        }
        executor.shutdown()
        while (!executor.isTerminated) {
            Thread.sleep(100)
        }

        val json = JSONObject(toReturn)
        return json.toString()
    }

    override fun getParams(params: MultiValueMap<String, String>) {
    }
}